explore/exploration.h

```diff
@@ -0,0 +1,207 @@
+  {
+    if (pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
+
+    size_t num_actions = pdf_last - pdf_first;
+    if (num_actions == 0)
+      return S_EXPLORATION_OK;
+
+	if (top_action >= num_actions)
+  	  top_action = (uint32_t)num_actions - 1;
+
+    float prob = epsilon / (float)num_actions;
+
+    for (It d = pdf_first; d != pdf_last; ++d)
+      *d = prob;
+
+    *(pdf_first + top_action) += 1.f - epsilon;
+
+    return S_EXPLORATION_OK;
+  }
+
+  template<typename It>
+  int generate_epsilon_greedy(float epsilon, uint32_t top_action, It pdf_first, It pdf_last)
+  {
+    typedef typename std::iterator_traits<It>::iterator_category pdf_category;
+    return generate_epsilon_greedy(epsilon, top_action, pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename InputIt, typename OutputIt>
+  int generate_softmax(float lambda, InputIt scores_begin, InputIt scores_last, std::input_iterator_tag scores_tag, OutputIt pdf_first, OutputIt pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    if (scores_last < scores_begin || pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
+
+    size_t num_actions_scores = scores_last - scores_begin;
+    size_t num_actions_pdf = pdf_last - pdf_first;
+
+    if (num_actions_scores != num_actions_pdf)
+    {
+      // fallback to the minimum
+      scores_last = scores_begin + ((std::min)(num_actions_scores, num_actions_pdf));
+      OutputIt pdf_new_last = pdf_first + ((std::min)(num_actions_scores, num_actions_pdf));
+
+      // zero out pdf
+      for (OutputIt d = pdf_new_last; d != pdf_last; ++d)
+        *d = 0;
+
+        pdf_last = pdf_new_last;
+    }
+
+    if (num_actions_scores == 0)
+      return S_EXPLORATION_OK;
+
+    float norm = 0.;
+    float max_score = *std::max_element(scores_begin, scores_last);
+
+    InputIt s = scores_begin;
+    for (OutputIt d = pdf_first; d != pdf_last && s != scores_last; ++d, ++s)
+    {
+      float prob = exp(lambda*(*s - max_score));
+      norm += prob;
+
+      *d = prob;
+    }
+
+    // normalize
+    for (OutputIt d = pdf_first; d != pdf_last; ++d)
+      *d /= norm;
+
+    return S_EXPLORATION_OK;
+  }
+
+  template<typename InputIt, typename OutputIt>
+  int generate_softmax(float lambda, InputIt scores_begin, InputIt scores_last, OutputIt pdf_first, OutputIt pdf_last)
+  {
+    typedef typename std::iterator_traits<InputIt>::iterator_category scores_category;
+    typedef typename std::iterator_traits<OutputIt>::iterator_category pdf_category;
+
+    return generate_softmax(lambda, scores_begin, scores_last, scores_category(), pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename InputIt, typename OutputIt>
+  int generate_bag(InputIt top_actions_begin, InputIt top_actions_last, std::input_iterator_tag top_actions_tag, OutputIt pdf_first, OutputIt pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    if (pdf_first == pdf_last || pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
```

Comment:
Here, pdf_first == pdf_last is E_EXPLORATION_BAD_RANGE while up above it is not.  This is inconsistent.

Reply:
done